// elevenlabs/dist/api/resources/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.audioNative = exports.models = exports.chapters = exports.projects = exports.voices = exports.user = exports.voiceGeneration = exports.speechToSpeech = exports.samples = exports.textToSoundEffects = exports.history = exports.workspace = exports.pronunciationDictionary = exports.dubbing = exports.textToSpeech = void 0;
exports.textToSpeech = __importStar(require("./textToSpeech"));
__exportStar(require("./textToSpeech/types"), exports);
exports.dubbing = __importStar(require("./dubbing"));
__exportStar(require("./dubbing/types"), exports);
exports.pronunciationDictionary = __importStar(require("./pronunciationDictionary"));
__exportStar(require("./pronunciationDictionary/types"), exports);
exports.workspace = __importStar(require("./workspace"));
__exportStar(require("./workspace/types"), exports);
exports.history = __importStar(require("./history"));
exports.textToSoundEffects = __importStar(require("./textToSoundEffects"));
exports.samples = __importStar(require("./samples"));
exports.speechToSpeech = __importStar(require("./speechToSpeech"));
exports.voiceGeneration = __importStar(require("./voiceGeneration"));
exports.user = __importStar(require("./user"));
exports.voices = __importStar(require("./voices"));
exports.projects = __importStar(require("./projects"));
exports.chapters = __importStar(require("./chapters"));
exports.models = __importStar(require("./models"));
exports.audioNative = __importStar(require("./audioNative"));
__exportStar(require("./history/client/requests"), exports);
__exportStar(require("./textToSoundEffects/client/requests"), exports);
__exportStar(require("./textToSpeech/client/requests"), exports);
__exportStar(require("./speechToSpeech/client/requests"), exports);
__exportStar(require("./voiceGeneration/client/requests"), exports);
__exportStar(require("./voices/client/requests"), exports);
__exportStar(require("./projects/client/requests"), exports);
__exportStar(require("./chapters/client/requests"), exports);
__exportStar(require("./dubbing/client/requests"), exports);
__exportStar(require("./audioNative/client/requests"), exports);
__exportStar(require("./pronunciationDictionary/client/requests"), exports);
__exportStar(require("./workspace/client/requests"), exports);
